source/adapters/opencl/kernel.hpp

```diff
@@ -0,0 +1,94 @@
+//===--------- kernel.hpp - OpenCL Adapter ---------------------------===//
+//
+// Copyright (C) 2023 Intel Corporation
+//
+// Part of the Unified-Runtime Project, under the Apache License v2.0 with LLVM
+// Exceptions. See LICENSE.TXT
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+#pragma once
+
+#include "common.hpp"
+#include "context.hpp"
+
+#include <vector>
+
+struct ur_kernel_handle_t_ {
+  using native_type = cl_kernel;
+  native_type Kernel;
+  ur_program_handle_t Program;
+  ur_context_handle_t Context;
+  std::atomic<uint32_t> RefCount = 0;
+
+  ur_kernel_handle_t_(native_type Kernel, ur_program_handle_t Program,
+                      ur_context_handle_t Context)
+      : Kernel(Kernel), Program(Program), Context(Context) {
+    RefCount = 1;
+    if (Program) {
```

Comment:
I think we can get away without null checking stuff like this in the constructors and destructors. It isn't optional so it'll already get checked by the validation layer, and if you're running without validation then you can't get too mad about hitting a segfault

Reply:
I changed that, I added it initially as `CreateWithNativeHandle` was implemented to accept it as nullptr, but now that I moved that logic to intel/llvm, this was removed.